### PR TITLE
[ENH] Integrate Autoformer model into V2 architecture

### LINKED ISSUE #2004: [ENH] Add Autoformer model
Referencing #1992 
The goal of this task is to assess the viability of integrating Autoformer into the current V2 interface and, if feasible, proceed with implementing the integration within the system.

CC: @fkiraly

### PR DESCRIPTION
**Pull request by harshsomankar123-tech**
### Description
This PR integrates the **Autoformer** model (based on the NeurIPS 2021 paper) into the new PyTorch-Forecasting V2 architecture. Autoformer implements a series-wise decomposition connection to progressively decompose time series into trend and seasonal components, and utilizes an Auto-Correlation mechanism to replace standard point-wise self-attention for long-term forecasting.

#### Reference Issues/PRs
 fix: #2004 
#### What does this implement/fix? Explain your changes.
This PR introduces the Autoformer architecture to the V2 models ecosystem. Specifically, it implements:

1. **Autoformer Layer Submodules (`pytorch_forecasting/models/autoformer/sub_modules.py`)**: 
   * `AutoCorrelation`: Fast Fourier Transforms (FFT) for periodicity discovery and time-delay aggregation.
   * `AutoCorrelationLayer`: Projection layer wrapper.
   * `series_decomp` & `moving_avg`: Custom trend-seasonal time series decomposition.
   * `Encoder` & `Decoder`: Progressive series-decomposition architecture.
   * Embeddings: `DataEmbedding_wo_pos`, `TokenEmbedding`, and temporal embeddings.
   * `my_Layernorm`: Specialized LayerNorm for seasonal components.

2. **Autoformer V2 Model (`pytorch_forecasting/models/autoformer/_autoformer_v2.py`)**: 
   * Inherits from `TslibBaseModel` to handle target values and continuous covariates in the V2 input dictionary.
   * Supports both standard regression losses and `QuantileLoss` predictions.
   * Implements channel dimension alignment to resolve potential broadcasting mismatches between target dimensions, covariates, and quantile channels during decoding.

3. **Wrapper Package & Registry (`pytorch_forecasting/models/autoformer/_autoformer_pkg_v2.py`)**: 
   * Specifies metadata tags (`exogenous`, `multivariate`, and `pred_int`) and parameter grids for dynamic discovery by the package registry.

#### What should a reviewer concentrate their feedback on?
Reviewers should primarily focus on:
* **Channel Dimension Alignment:** Specifically within `_autoformer_v2.py`, verifying the logic used to prevent broadcasting mismatches between targets, covariates, and quantile channels.
* **Quantile Loss Integration:** Ensuring the continuous covariates and target values are correctly handled alongside `QuantileLoss`.
* **V2 Input Dictionary Parsing:** Confirming that `TslibBaseModel` handles the dictionary routing to the encoder/decoder appropriately.

#### Did you add any tests for the change?
Yes, a comprehensive test suite was added in `tests/test_models/test_autoformer_v2.py`.
* **Custom Unit Tests:** 7/7 passed. Verifies model initialization, forward pass dimensions, and quantile predictions.
* **Automated Framework & Registry Tests:** 18/18 passed. Verifies checkpointing, serialization/reloading, prediction modes, and datamodule collation compatibility.

#### Any other comments?
All tests were verified locally using `.venv/bin/pytest tests/test_models/test_autoformer_v2.py` and `.venv/bin/pytest -k Autoformer`.

#### PR checklist
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.
- [x] Added/modified tests
- [x] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
  To run hooks independent of commit, execute `pre-commit run --all-files`

### CODE REVIEW

**Comment by harshsomankar123-tech** [comment]
@phoeenniixx @PranavBhatP @fkiraly Could you please take a look when you have some time?
Thanks!

File: pytorch_forecasting/models/autoformer/sub_modules.py
Comment (phoeenniixx) [inline pytorch_forecasting/models/autoformer/sub_modules.py:1]: I think this should move to `layers` module instead. Please look at the layers there, you might get some layers already implemented there, that can be used here

File: pytorch_forecasting/models/autoformer/sub_modules.py [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+import math
+from math import sqrt
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class my_Layernorm(nn.Module):
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/autoformer/sub_modules.py]: it should be `LayerNorm`? why add my?

File: pytorch_forecasting/models/autoformer/_autoformer_pkg_v2.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    @classmethod
+    def get_datamodule_cls(cls):
+        """Get the underlying DataModule class."""
+        from pytorch_forecasting.data.data_module import TslibDataModule
+
+        return TslibDataModule
+
+    @classmethod
+    def _get_test_datamodule_from(cls, trainer_kwargs):
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/autoformer/_autoformer_pkg_v2.py]: we dont need this method anymore, the `pkg` layer handles it under the hood

**Review by phoeenniixx — CHANGES_REQUESTED** [review]
Thanks!
I have added some suggestions based on a shallow review. Will do a more deeper review shortly

File: pytorch_forecasting/models/autoformer/sub_modules.py
Comment (harshsomankar123-tech) [reply pytorch_forecasting/models/autoformer/sub_modules.py:1]: @phoeenniixx PTAL 
Thanks!

Comment (phoeenniixx) [inline pytorch_forecasting/models/autoformer/sub_modules.py:1]: We dont need this file anymore? why not directly import in the main implementation file itself?

File: tests/test_models/test_autoformer_v2.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    )
+
+    assert model.hidden_size == hidden_size
+    assert model.n_heads == n_heads
+    assert model.e_layers == e_layers
+    assert model.d_layers == d_layers
+    assert model.n_quantiles is None
+
+
+def test_autoformer_forward(sample_dataset):
```
Comment (phoeenniixx) [inline tests/test_models/test_autoformer_v2.py]: I think the test framework is already testing this through [`test_integration`](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/tests/test_all_v2/_test_integration.py)?

File: pytorch_forecasting/models/autoformer/_autoformer_pkg_v2.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+from pytorch_forecasting.base._base_pkg import Base_pkg
+
+
+class Autoformer_pkg_v2(Base_pkg):
+    """Autoformer package container."""
+
+    _tags = {
+        "info:name": "Autoformer",
+        "info:compute": 2,
+        "authors": ["harshsomankar123-tech"],
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/autoformer/_autoformer_pkg_v2.py:15]: I think we are still missing some tags?

File: pytorch_forecasting/models/autoformer/_autoformer_pkg_v2.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        from pytorch_forecasting.metrics import SMAPE
+
+        params = [
+            # First set: smaller network params for fast testing
+            dict(
+                hidden_size=16,
+                n_heads=2,
+                e_layers=1,
+                d_layers=1,
+                d_ff=32,
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/autoformer/_autoformer_pkg_v2.py:51]: You can also add some `metrics` as loss here - to test if they work properly, That would save you from writing separate test for them

**Review by phoeenniixx — CHANGES_REQUESTED** [review]
Thanks!
Added a few comments, please have a look at them,
I think we should try to de duplicate the tests here and only tests those things which are not being tested by the test framework. 
See the tests for the test framework here: https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/tests/test_all_v2/test_all_estimators_v2.py

File: pytorch_forecasting/models/autoformer/sub_modules.py
Comment (harshsomankar123-tech) [reply pytorch_forecasting/models/autoformer/sub_modules.py:1]: Okay, I removed that file. Thanks!

**Comment by harshsomankar123-tech** [comment]
@phoeenniixx @fkiraly  Please Take a Look 
Thanks!
